design/internal-redirect_design.md

```diff
@@ -0,0 +1,179 @@
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+ name: myservice
+ namespace: prod
+spec:
+ virtualhost:
+   fqdn: www.example.com
+   routes:
+     - conditions:
+         - prefix: /echo
+       services:
+         - name: echo
+           port: 8080
+       internalRedirectPolicy:
+         maxInternalRedirects: 5
+         redirectResponseCodes: [ 301, 302, 303 ]
+         predicates:
+           safeCrossScheme: true
+           previousRoutes: true
+           allowedRouteNames: [ "other" ]
+         allowCrossSchemeRedirect: true
+     - name: other
+       conditions:
+         - prefix: /other
+       services:
+         - name: other
+           port: 8090
+```
+
+## Detailed Design
+
+Add a new type: `HTTPInternalRedirectPolicy` to define when the proxy should intercept and handle redirect responses internally.
+
+```Go
+// HTTPInternalRedirectPolicy defines when to handle redirect responses internally.
+type HTTPInternalRedirectPolicy struct {
+    // MaxInternalRedirects An internal redirect is not handled, unless the number of previous internal redirects that a downstream request has encountered is lower than this value.
+    // +optional
+    MaxInternalRedirects uint32 `json:"maxInternalRedirects,omitempty"`
+
+    // RedirectResponseCodes If unspecified, only 302 will be treated as internal redirect.
+    // Only 301, 302, 303, 307 and 308 are valid values.
+    // +optional
+    RedirectResponseCodes []uint32 `json:"redirectResponseCodes,omitempty"`
+
+    // Predicates are queried when an upstream response is deemed to trigger an internal redirect by all other criteria.
+    // +optional
+    Predicates HTTPInternalRedirectPredicate `json:"predicates,omitempty"`
+
+    // AllowCrossSchemeRedirect Allow internal redirect to follow a target URI with a different scheme than the value of x-forwarded-proto.
+    // +optional
+    AllowCrossSchemeRedirect bool `json:"allowCrossSchemeRedirect,omitempty"`
+}
+
+// HTTPInternalRedirectPredicate defines the predicate used to accept or reject an internal redirection.
+type HTTPInternalRedirectPredicate struct {
+    // An internal redirect predicate that checks the scheme between the downstream url and the redirect target url.
+    // +optional
+    SafeCrossScheme bool `json:"safeCrossScheme,omitempty"`
+
+    // An internal redirect predicate that rejects redirect targets that are pointing to a route that has been followed by a previous redirect from the current route.
+    // +optional
+    PreviousRoutes bool `json:"previousRoutes,omitempty"`
+
+    // AllowedRouteNames is the list of routes that’s allowed as redirect target
+    // by this predicate, identified by the route’s name.
+    //
+    // +optional
+    AllowedRouteNames []string `json:"allowedRouteNames,omitempty"`
+}
+```
+
+Update `Route` accordingly. 
+The `AllowedRouteNames` predicates declares a list of allowed route's name, so a `Name` field is required on `Route` to be able to use it.
```

Comment:
if this directly translates to this field: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-route-name

we will likely need to ensure uniqueness across all routes to make sure routes are configured correctly redirect predicates work correctly

Reply:
Yes, it translate to envoy route name field.

Is uniqueness required by envoy ? 

Regarding the internal redirect use, nothing prevent you to use the same name for multiple routes. For instance, if you have 2 routes that point to the same resource, you may use the same `Name`, and only declare a single entry in `AllowedRouteNames`.

Redirect predicates should work correctly, as its only job is to check if the route name is in the route allowed list.

Reply:
in a multi-tenant scenario where multiple teams/users are configuring HTTPProxies (or there are multiple HTTPProxies using the feature) we should make sure names don't conflict between users, that could generate unexpected results for people if route names are reused

Reply:
we'll need to think about this in the implementation section of this design/feature, something to keep track of here to make sure we provide a general enough API that is safe to use

Reply:
It seems the route name feature originates from debugging/troubleshooting use case ([see here](https://github.com/envoyproxy/envoy/issues/6364)) for printing to access log with `%ROUTE_NAME%` macro. Quickly browsing the code, I did not find any uniqueness requirements. 

Can allow list point even to a route name on different virtual host? 

In context of multi-tenant scenario, what would be natural scope for the allow list? For example: single namespace or e.g. root `HTTPProxy` (virtualhost) and its child proxies (possibly in other namespaces)?

Reply:
> In context of multi-tenant scenario, what would be natural scope for the allow list?

The problem with scoping the route name is that you limit the possible use cases. Team A may legitimately want to configure internal redirect for a route on Team B proxy, and configuring an internal redirect pointing on a different virtual host may also be very useful.

And the risk if we do not enforce route uniqueness is that in case of multiple teams provisioning the same proxy, team A configures an internal redirect on a route to handle a specific redirect response, and the team A service behind that route may also return a redirect response with the same status code an a location that matches a team B route with the same name. In such case, this other redirect is handled internally instead of being return to the client.

If the team A service does not returns redirect response pointing on Team B route (or use a different status code for internal redirects and other redirects), having duplicated routes on team B side will not have any impact for team A.

We have to find the right balance between risk mitigation and reducing the capability of this feature.

Reply:
Thinking more about it, enforcing route uniqueness may be the way to go.

It does not limit the possibility to use `Allow List` with redirects accros virtual hosts and accros namespaces (as long as teams agree on a route name), and the only drawback is that you may have to declare multiple routes instead of one in the allowed list.

Moreover, enforcing route name uniqueness will make the name more useful for debugging/troubleshooting routing issues.